File: app/src/test/java/com/github/goldy1992/mp3player/client/callbacks/metadata/MyMetadataCallbackTest.java
package com.github.goldy1992.mp3player.client.callbacks.metadata;

import android.os.Handler;
import android.support.v4.media.MediaMetadataCompat;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static android.support.v4.media.MediaMetadataCompat.METADATA_KEY_MEDIA_ID;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;


@RunWith(RobolectricTestRunner.class)
public class MyMetadataCallbackTest {

    private static MyMetadataCallback myMetadataCallback;

    private static MetadataListener metadataListener = mock(MetadataListener.class);

    @BeforeClass
    public static void setupClass() {
        Handler handler = mock(Handler.class);
        myMetadataCallback = new MyMetadataCallback(handler);
        myMetadataCallback.registerMetaDataListener(metadataListener);
    }

    @Test
    public void testNotifyListenerWithValidMetadata() {
        MediaMetadataCompat mediaMetadataCompat = new MediaMetadataCompat.Builder()
                .putString(METADATA_KEY_MEDIA_ID, "id")
                .build();
        myMetadataCallback.processCallback(mediaMetadataCompat);
        verify(metadataListener, times(1)).onMetadataChanged(mediaMetadataCompat);
    }

    @Test
    public void testNotifyListenerWithNullMetadataCompat() {
        myMetadataCallback.processCallback(null);
        verify(metadataListener, never()).onMetadataChanged(any());
    }

    @AfterClass
    public static void tearDownClass() {
        myMetadataCallback.removeMetaDataListener(metadataListener);
    }


}

File: app/src/test/java/com/github/goldy1992/mp3player/client/callbacks/metadata/MyMetadataCallbackTest.kt
package com.github.goldy1992.mp3player.client.callbacks.metadata

import android.os.Handler
import android.support.v4.media.MediaMetadataCompat
import org.junit.AfterClass
import org.junit.BeforeClass
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class MyMetadataCallbackTest {
    @Test
    fun testNotifyListenerWithValidMetadata() {
        val mediaMetadataCompat = MediaMetadataCompat.Builder()
                .putString(MediaMetadataCompat.METADATA_KEY_MEDIA_ID, "id")
                .build()
        myMetadataCallback!!.processCallback(mediaMetadataCompat)
        Mockito.verify(metadataListener, Mockito.times(1)).onMetadataChanged(mediaMetadataCompat)
    }

    @Test
    fun testNotifyListenerWithNullMetadataCompat() {
        myMetadataCallback!!.processCallback(null)
        Mockito.verify(metadataListener, Mockito.never()).onMetadataChanged(ArgumentMatchers.any())
    }

    companion object {
        private var myMetadataCallback: MyMetadataCallback? = null
        private val metadataListener = Mockito.mock(MetadataListener::class.java)
        @BeforeClass
        fun setupClass() {
            val handler = Mockito.mock(Handler::class.java)
            myMetadataCallback = MyMetadataCallback(handler)
            myMetadataCallback!!.registerMetaDataListener(metadataListener)
        }

        @AfterClass
        fun tearDownClass() {
            myMetadataCallback!!.removeMetaDataListener(metadataListener)
        }
    }
}